### ISSUE TITLE
Support Wayland

### ISSUE THREAD

**Issue by Salamandar** [linuxmint/Cinnamon#5201]
Well, you're basically the DE everyone wants supporting Wayland. You said it's on your roadmap, but with no precisions.
As Gnome Shell is supporting Wayland and you're based on it (a long time ago, but still), what are the challenges ? Wouldn't be a great news to do it for Cinnamon 3.0 ? :D

**Comment by lestcape** [comment]
Why take a risk and why lost time on this now? The wayland implementation of gnome shell is new. Is not better wait for a more stable release, how also occurs with Mint? Sure, if you use Arch or Debian testing, you will not agree with this, but well.

**Comment by Sunderland93** [comment]
Wayland in GNOME 3.20 is fully working, but it does not yet have support for Wacom tablets. I also expect to support Wayland in Cinnamon, and I hope it will be in 3.4-3.6.

**Comment by JosephMcc** [comment]
> Wayland in GNOME 3.20 is fully working, but it does not yet have support for Wacom tablets. I also expect to support Wayland in Cinnamon, and I hope it will be in 3.4-3.6.

If it's fully working then why does even Fedora not use it as the default?

**Comment by Salamandar** [comment]
> If it's fully working then why does even Fedora not use it as the default?
Because there are a lot of huge changes to do (distro/maintainer -side), and because some other projects (as Cinnamon) do not support Wayland.
I think Cinnamon not supporting Wayland is a brake for some distros to use Wayland by default.

**Comment by agur4ik** [comment]
> If it's fully working then why does even Fedora not use it as the default?
Because there are some things to implement but, as Fedora maintainers tell, it should not affect regular user and one may use Wayland session without noticing difference in stability and functionality if no specific unimplemented features are needed. I have tried Gnome 3.20 on Arch, and I can assure that it completely works with no flaws so far. But Gnome is such a bad-suited for desktop laggy shit.  BTW, GDM already uses Wayland by default, at least it does on 3.18 in Arch, dunno what about other distros. 

> I think Cinnamon not supporting Wayland is a brake for some distros to use Wayland by default
You are completely wrong, distros themselves cannot use by default either Wayland or X for entire system, it is quite specific to DEs. If, for example, Fedora wants to switch their default Gnome session to Wayland  there's nothing in Cinnamon or Xfce or Mate etc. to prevent them from doing so, yet Cinnamon, Mate and Xfce would be available only as X sessions. BTW, almost every GTK3 app is fully functional in Wayland, I have tested even Nemo in Weston with xwayland extension disabled and it worked with desktop handling turned off (otherwise, it segfaulted, which is completely excusable for software not designed to run in Wayland), with GTK update to 3.20 even Drag&Drop started working in Nemo, which was kind of surprising for me. 

The main problem, I guess, is GTK+ and Wayland versions in Linux Mint and LMDE. Such important feature as Drag&Drop is present in GTK since 3.20 and in quite recent version of Wayland itself (I don't remember the exact version). GTK in Ubuntu 16.04 (base for Mint 18.0) is 3.18 version, in Debian Jessie (LMDE base) it is even older, 3.14, I guess. For now adoption of Cinnamon to Wayland would make sense only if Mint was based on some bleeding edge or rolling release distro (at least not on Debian Stable with its necroversions of software and not on LTS Ubuntu, which becomes outdated after a half year after release). In cold reality Cinnamon is doomed to run and be developed on Ubuntu 16.04 and Debian 8 (if there are no plans for LMDE 3 using testing branch) based distros for about 2 and 1 years respectively. 

P.S.: sorry for my English.

**Comment by Anti-Ultimate** [comment]
I'd like to bring my two cents into this discussion

First of all, it'd make very much sense to start implemementing Wayland support into Muffin right now. With the current Linux Mint release cycle, of course it would not be implemented inside Linux Mint, but implementing Wayland is not going to be a "it's done in the next release" thing. Developers will put it into Linux Mint once it's stable, but that doesn't stop anyone on this project to start forking Muffin and implement it.(And later make a pull request).

All that can be said is just that once Cinnamon gets support for Wayland, we'll see more and more support and distros switching. Considering Linux Mint is basically the most used Linux distro that is out there, it's the one to give directions. If Mint won't change to Wayland, other distros will be afraid too.

**Comment by mtwebster** [comment]
I think you give Mint and Cinnamon too much credit here - I seriously doubt any other distros are waiting for us to make a move one way or the other.

Right now for us (for me at least especially) it comes down to manpower vs payoff - if we dictate from On High that we will support wayland in some future release, someone needs to actually work on that.  This would very likely prevent said person from participating much in other development, which right now would probably have a heavy impact on ongoing releases and maintenance.  With only a handful of developers we have currently, this is a tough prospect.  I will say though that this last cycle we've seen a number of new folks getting involved, which is great, and might make something like this possible eventually.

Another point I want to make, which gets forgotten a lot, is that Cinnamon is a DE for mint.  It was never developed to be a freestanding entity (Like Mate or XFCE or what-have-you) - as such, we develop primarily with Mint users in mind.  Most Mint users don't care or have any awareness of Wayland or X or whatever, until something breaks and they read stuff online.  We try to present something cohesive to our users, a complete package - desktop, apps, whatever.  And from that point of view, wayland is currently irrelevant for us.  We want to give you an experience, not an ISO with upstream packages and no fore-thought.  We're definitely not focused on the bleeding edge of development, unless it overlaps what we think our users want and is reasonably stable.

This is not to say we don't care about users running Cinnamon on other distros, but their desires or expectations are probably very different from the average Mint user, as well as their experience level or willingness to tinker with stuff.

Sooner or later this will have to be addressed - if for no other reason than the fact that libraries we rely on are steadily having their X11 support deprecated, broken, or removed.  We're being inexorably pushed in this direction (just like systemd).  I'd like nothing better than to throw everything away and start from ground zero and cast off the oligarchy that dictates the reinvention of everything every few years (or six months for GTK's case,) but I'm neither smart enough, nor do I have 23 hours a day to dedicate to such a crusade :)

**Comment by leigh123linux** [comment]
I'm not interested in wayland for fedora cinnamon.

**Comment by cjermain** [comment]
@mtwebster, I think many Mint users will be interested in Wayland for multi-dpi support. Many new laptops come with HiDPI screens -- even the cheap ones. Although they can usually be patched (with gsettings and xrandr) for a single screen, there are significant problems when connecting to other DPI monitors or TVs. Wayland promises to solve mulit-dpi problems. I suspect this will be the main intensive for Wayland, and is why I'm personally paying attention.

**Comment by cdlans** [comment]
I'm interested in Wayland for Cinnamon, for the same reasons as @cjermain and also I hear Wayland would be able to scale the desktop by a factor of 1.5 on my HD screen.

**Comment by TheChuckster** [comment]
Also hoping for mixed-dpi Wayland support in Mint/Cinnamon, keep up the good work!

**Comment by BloodShura** [comment]
Interested!

**Comment by filakhtov** [comment]
Fedora is close to getting Wayland by default, there are few interesting related links:

* [Gnome Blog: WAYLAND BY DEFAULT IN FEDORA 25?](https://blogs.gnome.org/mcatanzaro/2016/08/25/wayland-by-default-in-fedora-25/)
* [Fedoraproject wiki: Changes/WaylandByDefault](https://fedoraproject.org/wiki/Changes/WaylandByDefault)

Although it still can change, they are aiming at achieving that. I own XPS15 with Full HD screen and it is not possible to get solid setup with external display, nor good enough font/UI elements scale on laptop screen. x1.5 multiplier would be just great which is one of the Wayland features as well.

**Comment by leigh123linux** [comment]
@filakhtov 

Wayland breaks a lot of apps in fedora, I'm sure FESCO will regret their dumb decision  to use it for F25

**Comment by filakhtov** [comment]
@leigh123linux, could you please point me out into problematic/broken apps in Fedora with Wayland? I'm about to build Gentoo with Gnome for my MacBook (since Cinnamon is a pain w/ multi-display setup) and would like to know what and where to check. Thanks in advance and sorry for off-topic.

**Comment by leigh123linux** [comment]
@filakhtov 

A lot of the media players won't work properly 


https://bugzilla.redhat.com/show_bug.cgi?id=1373217

and there is totally broken video acceleration


https://bugzilla.redhat.com/show_bug.cgi?id=1305699

https://bugzilla.redhat.com/show_bug.cgi?id=1373217

**Comment by filakhtov** [comment]
@leigh123linux, thank you, sir! I will keep an eye on these during my experiments!

**Comment by 0x414A** [comment]
+1 on Wayland support for multi-DPI setups.

@cdlans Wayland does indeed allow you to have 1.5x scaling :smile: . I was using GNOME Wayland and it worked beautifully for my laptop (which is HiDPI at 2x scaling), and my external monitor (1.5x scaling). No more overly large or overly small elements on one monitor lol. 

But yeah I prefer Cinnamon, so it would be fantastic to see it run as a Wayland session.

**Comment by cdlans** [comment]
Hi @0x414A, on which distro were you using GNOME Wayland? Can you recommend it? I want to give kubuntu a try as soon as I have some time.

**Comment by 0x414A** [comment]
@cdlans Honestly I just set it up on Arch. [Ubuntu Gnome](https://wiki.ubuntu.com/XenialXerus/ReleaseNotes/UbuntuGNOME) seems to have the wayland session available, though, if you want something set up out of the box.

**Comment by cdlans** [comment]
@0x414A I'm kind of bound to Ubuntu since Android Studio only "supports" Ubuntu. I did a quick test of kubuntu (Live USB). I thought it supported wayland out of the box, but it doesn't (yet). However I could scale fonts to a factor of 1.5 (or any decimal fraction). But icons, buttons and other GUI elements only scale to integer factors.

**Comment by leigh123linux** [comment]
Can all the +1's read this bug report and honestly say they still want wayland after reading it!

**Comment by filakhtov** [comment]
@leigh123linux: I read it, and I honestly I still want Wayland. I think there is some misunderstanding in how linux auth stack is working, i. e. misusing `sudo`. When it comes to graphical stack we've got `polkit` as a proper mechanism for handling super privileges and I think this is what should be done in this case as well. So, `pkexec`. Just my 5¢ on the topic.

**Comment by ManIVIctorious** [comment]
@leigh123linux i have to admit i don't see your point here. The main purpose of wayland isn't the ability of multi dpi, nor the tear free rendering.

It is invented to (massively) increase the system's security and running applications as root is a really bad idea. For the mentioned problems systems like polkit and logind were invented to gian privileged rights when needed. These issues would have to be adressed by the specific programs and therefore its no wayland issue at all

**Comment by Salamandar** [comment]
@leigh123linux Same as the other answers. 
And yes, the Wayland move will be a difficult move. But that doesn't mean you should not do it. Everyone is doing this port (apart from Unity/Mir), and it's a logical idea to do it. More, the Cinnamon devs know they will have to do it someday. Better start now ! ;)

**Comment by lestcape** [comment]
Hey, people, the move to Wayland will occurs (thats clear). This is just a question of when will be the best time, when this will create bugs and problems or when we have more documentation, more application running under Wayland, more problem fixed, more code with examples, more optimizations... For me it's clear when will be the best time and when not.

**Comment by fpqc** [comment]
@lestcape If the mint team wants to start a wayland fork, I'm willing to package and maintain it on the AUR for Arch (call it cinnamon-wayland-git) (that way you can start getting bug-reports and stuff so by the time you ship for Mint, it will be stable). 

Let me know =).

**Comment by leigh123linux** [comment]
Screw wayland, they can fix their own f**k up https://bugzilla.redhat.com/show_bug.cgi?id=1403429

**Comment by Salamandar** [comment]
@leigh123linux Again, what's your point ? That any new HW/SW has bugs at the beginning ? Thanks, but we already knew that.

**Comment by leigh123linux** [comment]
@Salamandar My point is Linuxmint won't port cinnamon to wayland till it's mature and issue like this have been fixed.

**Comment by Anti-Ultimate** [comment]
According to that bug report, it doesn't like like GNOME/Wayland could do much about it. The program expects a display.

Anyways, I think Wayland just needs unified display settings. Things like changing Display Resolution, Mouse Acceleration or exclusive fullscreen (VERY important for games) should be unified in every Wayland compositor so we don't have different code for each Window Manager. X11 is old, but you can't say it was bad to have everything made for X11 to work with any DE. OBS or SimpleScreenRecorder will need to implemented code for every Wayland compositor until something like the above example is established.

**Comment by fpqc** [comment]
@Anti-Ultimate  The problem with X was that there were  always about 50 different places to set settings, many of which conflicted with, for example, the DE or the DM, especially, for example, keymapping and display settings.

Hopefully as the DEs move over to Wayland proper, they will take over global key mapping and configuration of display settings, etc. Without the xkbmap and xrandr-based ugly hacks, the DE-writers will actually implement these things instead of pointing people to using unstable workarounds.

**Comment by Vistaus** [comment]
@leigh123linux Have you ever looked at the problems people are experiencing with Xorg and how many people are reporting problems? It's not much better in that regard even though it's supposed to be more mature.

**Comment by llamaha** [comment]
I also suffer from the multi-dpi issue on my QHD+ Dell XPS, it would have been much easier to go with FHD and avoid all of these issues but now I think I have to move to Ubuntu Gnome for multi-dpi support as it's a necessary feature for work... 

Personally I think that it would be better to deprecate either MATE or KDE and start working towards a Wayland release so that new laptops are HiDPI compatible, especially laptops like the XPS which has official driver support from the manufacturer.

**Comment by JosephMcc** [comment]
> Personally I think that it would be better to deprecate either MATE or KDE and start working towards a Wayland release so that new laptops are HiDPI compatible, especially laptops like the XPS which has official driver support from the manufacturer.

To be fair, Cinnamon is HiDPI compatible. The issue comes from mixing HiDPI and non-HiDPI monitors.

**Comment by Salamandar** [comment]
Cinnamon only has a 2× multiplier. A finer setting would be allowed by Wayland ;)

**Comment by JosephMcc** [comment]
> Cinnamon only has a 2× multiplier. A finer setting would be allowed by Wayland ;)

Says who? This is decided by the toolkit used. Gtk only allows an integer multiplier.

**Comment by clefebvre** [comment]
GTK allows 1x, 2x and 3x integer ratios afaik. I don't think it's an artificial constraint though, I think their hands are tied by other constraints which they have to deal with (I heard something about fonts being one of them).

Going back to the main topic, it's not our role to decide whether Wayland is going to be a success or not. If there's a demand for it coming from distributions we need to accommodate it and support that technology. That said, this would be a huge WIP and there's no internal need for it here in the scope of Mint, so work on this would be driven by the distributions which need that support.

In other words external PRs bringing support for Wayland, even partially, could be considered for merging and I think we should welcome them when/if they come, but it's very unlikely Wayland support will land in Mint's Roadmap itself any time soon.

Things can change of course, but it's not on the horizon right now anyway.

**Comment by fpqc** [comment]
@clefebvre GNOME3 is going to be deprecating the X versions of some of their stuff over the next 2 or 3 years according to Ikey from Solus, so unless you plan to completely fork like with MATE, you will eventually be forced by events to make the move.

**Comment by i23098** [comment]
It really boils down to what Cinnamon wants to be... If it is "just" a DE for Linux Mint, that is also available (but not really supported) to other distributions, then, it's ok to don't support Wayland any time soon...

But I think it is much more than that... It's an awesome DE, used in many distros, Cinnamon is great :) So, adding support to Wayland sooner rather than later would be really good.

Now, maybe the non-Mint users (like myself) should be making pressure on ours distros side to help Cinnamon team to support Wayland. Fedora moved to Wayland as default, Fedora Cinnamon Spin Team, should work with Cinnamon Team...

**Comment by fpqc** [comment]
Fedora also doesn't support the nvidia blob drivers (not Wayland compatible), so they don't really need X support.

**Comment by i23098** [comment]
Fedora + nvidia is a bit offside to this conversation... (but work is being done - https://blogs.gnome.org/uraeus/2016/11/01/discrete-graphics-and-fedora-workstation-25/)

**Comment by fpqc** [comment]
@i23098 That's not relevant.  That's about the dual-gpu setup.

**Comment by i23098** [comment]
@fpqc as I said, it's a big offside to discuss here... but if you read the post it says that "A big part of this story is also that we are trying to make it easier to run the NVidia binary driver on Fedora in general"

Fedora is moving to Wayland, that is a fact, with nvidia blob drivers or not... Other distros will probably start moving to Wayland, so, I would like to have Cinnamon on Wayland.

**Comment by cjermain** [comment]
@clefebvre, I'm interested to hear your thoughts on support in Cinnamon for multiple monitors with different DPIs (specifically relevant as Wayland offers different options over GTK). Is this a feature that you see in the pipeline for Cinnamon?

**Comment by DiJu519** [comment]
Does Ubuntu dropping Mir[and unity] in favour of Wayland/Gnome make this Feature Request any more compelling? 

> Drop MIR/Unity for Wayland/Gnome (351 weight)
> Release/GA Unity 8 (15 weight)
> Easily, the most heavily requested, major change in this thread was for Ubuntu to drop MIR/Unity in favor of Wayland/Gnome.  And that's exactly what Mark Shuttleworth announced in an Ubuntu Insights post here today.  There were a healthy handful of Unity 8 fans, calling for its GA, and more than a few HackerNews comments lamenting the end of Unity in this thread.
> 
> http://blog.dustinkirkland.com/2017/04/thank-you-note-to-hackernews.html`

**Comment by leigh123linux** [comment]
@DiJu519 

Wayland adding Nvidia support would be more compelling than so lame fork dying.

**Comment by thendrix** [comment]
Is there more interest in this on cinnamon side now GNOME 3.24 mainline has Wayland support for Nvidia via EGLStreams, or you guys waiting for new unified API or the next LTS sync?

**Comment by leigh123linux** [comment]
@thendrix 

Last time I tested this so claimed support it wasn't working.

> GNOME 3.24 mainline has Wayland support for Nvidia via EGLStreams

The archlinux users seem to be having similar issues

https://bugs.archlinux.org/task/53284

**Comment by kellerkindt** [comment]
Any news to this? Since ubuntu might ship 17.10 with wayland and is planning to ship it from 18.04 onward...

**Comment by NikoKrause** [comment]
Ubuntu 18.04 will ship with xorg.

**Comment by clefebvre** [comment]
Many thanks for contributing to Cinnamon. Your suggestion was reviewed.

For more information on our workflow and feature requests, read https://linuxmint-troubleshooting-guide.readthedocs.io/en/latest/faq.html.
